pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -155,19 +156,24 @@ def _call_children_scripts(self):
         os.environ["NODE_RANK"] = str(self.cluster_environment.node_rank())
         os.environ["LOCAL_RANK"] = str(self.cluster_environment.local_rank())
 
-        # when user is using hydra find the absolute path
-        path_lib = os.path.abspath if not _HYDRA_AVAILABLE else to_absolute_path
-
-        # pull out the commands used to run the script and resolve the abs file path
-        command = sys.argv
-        try:
-            full_path = path_lib(command[0])
-        except Exception:
-            full_path = os.path.abspath(command[0])
-
-        command[0] = full_path
-        # use the same python interpreter and actually running
-        command = [sys.executable] + command
+        # Check if the current calling command looked like `python a/b/c.py` or `python -m a.b.c`
+        # See https://docs.python.org/3/reference/import.html#main-spec
+        if __main__.__spec__ is None:  # Script called as `python a/b/c.py`
+            # when user is using hydra find the absolute path
+            path_lib = os.path.abspath if not _HYDRA_AVAILABLE else to_absolute_path
+
+            # pull out the commands used to run the script and resolve the abs file path
+            command = sys.argv
+            try:
+                full_path = path_lib(command[0])
+            except Exception:
+                full_path = os.path.abspath(command[0])
+
+            command[0] = full_path
+            # use the same python interpreter and actually running
+            command = [sys.executable] + command
+        else:  # Script called as `python -m a.b.c`
+            command = [sys.executable, "-m", __main__.__spec__.name] + sys.argv[1:]
```

Comment:
don't we need to find the hydra path here as well?

Reply:
I haven't worked with Hydra, and don't know why it needs the original path, or how to resolve that requirement here. If you could suggest how to make hydra work with this case, I can add that to the pull request.

Alternatively, I can add `assert not _HYDRA_AVAILABLE` here and document that Hydra is not supported with "python -m" syntax. That would still be an improvement over the current state of things.

Please let me know how to move forward.

Reply:
Never mind, I read about it, and the reason that we needed Hydra's to_absolute_path for the `python a/b.py` case is that hydra copies the file to a different location and runs the copied file instead of the main one, so `os.path.abspath` would have returned the wrong result.

But the module name `a.b` in this case isn't actually changed. And the new process will be able to find the module even using the original name because we're calling the subprocess with `cwd=get_original_cwd()`.

Just to be sure, I installed Hydra and tested with it, it all works.